revocation/crl/cache/file.go

```diff
@@ -0,0 +1,250 @@
+		return err
+	}
+
+	// save to temp file
+	tempFile, err := os.CreateTemp("", tempFileName)
+	if err != nil {
+		return err
+	}
+	if err := saveTar(tempFile, bundle); err != nil {
+		return err
+	}
+	tempFile.Close()
+
+	// rename is atomic on UNIX-like platforms
+	return os.Rename(tempFile.Name(), filepath.Join(c.root, fileName(uri)))
+}
+
+// fileName returns the file name of the CRL bundle tarball
+func fileName(url string) string {
+	return hashURL(url) + ".tar"
+}
+
+// hashURL hashes the URL with SHA256 and returns the hex-encoded result
+func hashURL(url string) string {
+	hash := sha256.Sum256([]byte(url))
+	return hex.EncodeToString(hash[:])
+}
+
+// parseBundleFromTar parses the CRL blob from a tarball
+//
+// The tarball should contain two files:
+// - base.crl: the base CRL in DER format
+// - metadata.json: the metadata of the CRL
+//
+// example of metadata.json:
+//
+//	{
+//	  "base.crl": {
+//	    "url": "https://example.com/base.crl"
+//	  },
+//	  "createAt": "2024-07-20T00:00:00Z"
+//	}
+func parseBundleFromTar(data io.Reader) (*Bundle, error) {
+	bundle := &Bundle{}
+
+	// parse the tarball
+	tar := tar.NewReader(data)
+	for {
+		header, err := tar.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, &BrokenFileError{
+				Err: fmt.Errorf("failed to read tarball: %w", err),
+			}
+		}
+
+		switch header.Name {
+		case PathBaseCRL:
+			// parse base.crl
+			data, err := io.ReadAll(tar)
+			if err != nil {
+				return nil, err
+			}
+
+			var baseCRL *x509.RevocationList
+			baseCRL, err = x509.ParseRevocationList(data)
+			if err != nil {
+				return nil, &BrokenFileError{
+					Err: fmt.Errorf("failed to parse base CRL from tarball: %w", err),
+				}
+			}
+			bundle.BaseCRL = baseCRL
+		case PathMetadata:
+			// parse metadata
+			var metadata Metadata
+			if err := json.NewDecoder(tar).Decode(&metadata); err != nil {
+				return nil, &BrokenFileError{
+					Err: fmt.Errorf("failed to parse CRL metadata from tarball: %w", err),
+				}
+			}
+			bundle.Metadata = metadata
+		}
+	}
+	if err := bundle.Validate(); err != nil {
+		return nil, err
+	}
+
+	return bundle, nil
+}
+
+// SaveAsTar saves the CRL blob as a tarball, including the base CRL and
+// metadata
+//
+// The tarball should contain two files:
+// - base.crl: the base CRL in DER format
+// - metadata.json: the metadata of the CRL
+//
+// example of metadata.json:
+//
+//	{
+//	  "base.crl": {
+//	    "url": "https://example.com/base.crl"
+//	  },
+//	  "createAt": "2024-06-30T00:00:00Z"
+//	}
+func saveTar(w io.Writer, bundle *Bundle) (err error) {
+	tarWriter := tar.NewWriter(w)
+	defer func() {
+		if cerr := tarWriter.Close(); cerr != nil && err == nil {
+			err = cerr
+		}
+	}()
+
+	// Add base.crl
+	if err := addToTar(PathBaseCRL, bundle.BaseCRL.Raw, bundle.Metadata.CreatedAt, tarWriter); err != nil {
+		return err
+	}
+
+	// Add metadata.json
+	metadataBytes, err := json.Marshal(bundle.Metadata)
```

Comment:
QQ: shall we add `NextUpdate` into bundle metadate to help cache refresh?

Reply:
Yes, we need a discussion on it, but let's wait for CRL #214 to be merged first.

Reply:
Added `NextUpdate`